sds_in_a_box/SDSCode/upload_api.py

```diff
@@ -0,0 +1,117 @@
+        if pattern[field] == '*':
+            file_dictionary[field] = split_filename[i]
+        elif pattern[field] == split_filename[i]:
+            file_dictionary[field] = split_filename[i]
+        else:
+            return None
+        i += 1
+    
+    return file_dictionary
+
+def _generate_signed_upload_url(filename, tags={}):
+    '''
+    Based on a given filename, this function will open up a presigned url into the correct location on the SDS storage bucket
+    
+    :param filename: Required.  A string representing the name of the object to upload.  
+    :param tags: Optional.  A dictionary object of key:value pairs that will be stored in the S3 object metadata.  
+    
+    :return: None if the filename does not match mission naming conventions.  Otherwise, a URL string.  
+     '''
+    filetypes = _load_allowed_filenames()
+    for filetype in filetypes:
+        path_to_upload_file = filetype['path']
+        metadata = _check_for_matching_filetype(filetype['pattern'], filename)
+        if metadata is not None:
+            break
+        
+    if metadata is None:
+        logger.info(f"Found no matching file types to index this file against.")
+        return None
+    
+    bucket_name = os.environ["S3_BUCKET"]
+    url = boto3.client('s3').generate_presigned_url(ClientMethod='put_object', 
+                                                    Params={'Bucket': bucket_name[5:], 
+                                                    'Key': path_to_upload_file+filename, 
+                                                    'Metadata': tags}, 
+                                                    ExpiresIn=3600)
+    
+    return url
+
+def lambda_handler(event, context):
+    '''
+    The entry point to the upload API lambda.  
+    This function returns an S3 signed-URL based on the input filename, which the user can 
+    then use to load a file onto the SDS. 
+    
+    :param event: Dictionary object.  
+                  Specifically only requires event['queryStringParameters']['filename'] be present. 
+                  User-specified key:value pairs can also exist in the 'queryStringParameters', storing these pairs as object metadata.  
+    :param context: Unused
+    
+    :return: If all checks are successful, this returns a pre-signed url where users can upload a data file to the SDS.  
+    '''
+    
+    if 'filename' not in event['queryStringParameters']:
+        return {
+            'statusCode': 400,
+            'body': json.dumps("Please specify a filename to upload")
+        }
+        
+    filename = event['queryStringParameters']['filename']
+    url = _generate_signed_upload_url(filename, tags=event['queryStringParameters'])
+    
+    if url is None:
+        return {
+            'statusCode': 400,
+            'body': json.dumps("A pre-signed URL could not be generated. Please ensure that the file name matches mission file naming conventions.")
+        }
+    
+        
+    return {
```

Comment:
what is supposed to happen with the returned url? Is that supposed lead to a page where you can upload files?

Reply:
Pretty much, though its not a drag and drop page or anything.  It's a spot where a user can do a http PUT request to upload files.  

For example:
 `with open("file.txt", 'rb') as object_file:
        object_text = object_file.read()
    response = requests.put(upload_url, data=object_text)`

Reply:
thanks! I think you made this change in the original larger PR, but the config.json needs to have a "/" added to the end of the path (or add it when you combine the path and filename. When I upload a file it's currently adding l0 as a prefix to the filename instead of it being the directory that the file is contained in.

Reply:
Ah yes now that you mention it, I did make changes to the config.json!  I'll be sure to put that in this commit.